csi.proto

```diff
@@ -502,6 +556,213 @@ message ControllerServiceCapability {
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. This field is REQUIRED.
+  string timestamp = 4;
+
+  // The status of a snapshot. A snapshot could have the following
+  // statuses:
+  // CREATING - A snapshot is being cut.
+  // UPLOADING - A snapshot is cut and is now being uploaded.
+  // READY - A snapshot is ready for use.
+  // ERROR - Error occurred when creating or uploading the snapshot.
+  // Some cloud providers will upload the snapshot to a location in
+  // the cloud (i.e., an object store) after the snapshot is cut.
+  // Uploading is a long process that could take hours. At this phase,
+  // since the snapshot is already cut, a `thaw` operation can be
+  // performed so application can start running again (assuming a
+  // `freeze` operation was done before taking the snapshot). The
+  // status of the snapshot will become `READY` after the upload is
+  // complete.
+  // For cloud providers and storage systems that don't have the
+  // uploading process, the status should be `READY` after `CREATING`.
+  // `thaw` can be done when the status is `READY` in this case.
+  SnapshotStatus status = 5;
+}
+
+// The status of a snapshot. This parameter provides guidance to the CO
+// on what action can be taken in the process of snapshotting. Based on
+// this information, CO can issue repeated (idemponent) calls to
+// CreateSnapshot, monitor the response, and make decisions.
+message SnapshotStatus {
+  enum Type {
+     UNKNOWN = 0;
+     // A snapshot is being cut.
+     CREATING = 1;
+     // A snapshot is cut and is now being uploaded.
+     UPLOADING = 2;
+     // A snapshot is ready for use.
+     READY = 3;
+     // Error occurred when creating or uploading the snapshot.
+     ERROR = 4;
+  }
+  Type type = 1;
+}
+message DeleteSnapshotRequest {
+  // The ID of the snapshot to be deleted.
+  // This field is REQUIRED.
+  string snapshot_id = 1;
+
+  // Secrets required by plugin to complete snapshot deletion request.
+  // A secret is a string to string map where the key identifies the
+  // name of the secret (e.g. "username" or "password"), and the value
+  // contains the secret data (e.g. "bob" or "abc123").
+  // Each key MUST consist of alphanumeric characters, '-', '_' or '.'.
+  // Each value MUST contain a valid string. An SP MAY choose to accept
+  // binary (non-string) data by using a binary-to-text encoding scheme,
+  // like base64.
+  // An SP SHALL advertise the requirements for required secret keys and
+  // values in documentation.
+  // CO SHALL permit passing through the required secrets.
+  // A CO MAY pass the same secrets to all RPCs, therefore the keys for
+  // all unique secrets that an SP expects must be unique across all CSI
+  // operations.
+  // This information is sensitive and MUST be treated as such (not
```

Comment:
both for the create and the delete it seems a little weird that a user is going to give their username and password (aka sensitive data) to a third party vendor plugin. What prevents a malicious plugin from passing the user-name and secret to other entities or caching the username/password at locations that aren't secure?

There is no relation between the volume user (end user thats owning the data and the app) and the volume plugin that's facilitating the allocation of storage.

Passing in secrets to the volume provider seems like a very insecure thing to do and an additional burden on the volume plugin to not misuse this data.